### DISCUSSION TITLE
Writing TrueType instructions to a glif

### DISCUSSION THREAD

**Discussion by psb1558** [fonttools/fonttools#2991] [Q&A]
I'd like to add TT instructions to a glif in a ufo using ufoLib. As far as I can see (so far), the way to do this is to get a glyphSet, get the contents of the glif with glyphSet.readGlyph() (supplying an instance of a very simple glyph object class), add the TrueType code to the lib, and use writeGlyph to put everything back:

```
gs = ufo.getGlyphSet()
gs.readGlyph("A", glyphObject=glyph_object, ...)
go.lib["public.truetype.instructions"]["assembly"] = asm
gs.writeGlyph("A", glyphObject=glyph_object, ...)
```

My problem is where I've left the ellipses above. To avoid losing the outline data, I've got to supply a pointPen instance for readGlyph and a function that takes a pointPen as a parameter for writeGlyph. But the fontTools docs are vague about what goes into that function ("it has to call the proper PointPen methods to transfer the outline to the .glif file"). I can't figure out what that function should look like.

So here are my questions:

- Am I right about what needs to be done here--that there seems to be no more straightforward way to write TT instructions to the .glif than to read *all* the data from it and then write it all back?
- If so, can anyone point to an example of how this is done--especially storing and restoring the outline data?

I'll be grateful for any guidance.

**Comment by psb1558** [comment]
I see now that I can simply add the instructions with etree:

```
from fontTools import ufoLib
from fontTools.misc import plistlib
from lxml import etree

ufo_name = "Elstob-Regular.ufo"
u = ufoLib.UFOWriter(ufo_name)
gs = u.getGlyphSet()
gxml = etree.fromstring(gs.getGLIF("A"))
lib_list = gxml.xpath("/glyph/lib")
if len(lib_list) == 0:
    lib_el = etree.SubElement(gxml, "lib")
else:
    lib_el = lib_list[0]
plist = plistlib.fromtree(lib_el)
if not "public.truetype.instructions" in plist:
    plist["public.truetype.instructions"] = {}
plist["public.truetype.instructions"]["formatVersion"] = "1"
plist["public.truetype.instructions"]["assembly"] = "PUSHB_1[ ] 1 POP[ ]"
etree.strip_elements(lib_el, "dict")
lib_el.append(plistlib.totree(plist))
f = open(ufo_name + "/glyphs/" + gs.contents["A"], 'w')
f.write(etree.tostring(gxml, pretty_print=True).decode())
f.close()
u.close()
```
I'll do it this way unless someone can suggest a better (easier or more efficient) way of doing it.

**Comment by justvanrossum** [comment]
You'd be much beter off using a higher level libary such as ufoLib2.

**Reply by psb1558** [reply]
Thanks -- that looks like just the thing.
